rlp/encode.go

```diff
@@ -397,22 +399,93 @@ func writeBigIntNoPtr(val reflect.Value, w *encbuf) error {
 }
 
 func writeBigInt(i *big.Int, w *encbuf) error {
-	if cmp := i.Cmp(big0); cmp == -1 {
+	switch i.Sign() {
+	case -1:
 		return fmt.Errorf("rlp: cannot encode negative *big.Int")
-	} else if cmp == 0 {
+	case 0:
 		w.str = append(w.str, 0x80)
-	} else {
-		w.encodeString(i.Bytes())
+	default:
+		bitlen := i.BitLen()
+		if bitlen <= 64 {
+			w.encodeUint(uint64(i.Bits()[0]))
+		} else {
+			writeBigIntBits(bitlen, i.Bits(), w)
+		}
 	}
 	return nil
 }
 
+// wordBytes is the number of bytes in a big.Word
+const wordBytes = (32 << (uint64(^big.Word(0)) >> 63)) / 8
+
+// writeBigIntBits encodes big.Int values > 2^64.
+func writeBigIntBits(bitlen int, words []big.Word, w *encbuf) {
+	length := ((bitlen + 7) & -8) >> 3
```

Comment:
could you explain this formula a bit? It's not obvious what it means

Reply:
done